source/_drafts/Feature-Spotlight-Middleware.md

```diff
@@ -0,0 +1,187 @@
+
+        // call next with an error to signal you cannot handle the source
+        next(new Error('Source is not an HLS source'));
+      } else {
+
+        // in here we know we're playing back an HLS source.
+        // We don't want to do anything special for it, so, pass along the source along with a null.
+        next(null, srcObj);
+      }
+    },
+
+    // this method gets called on the tech and then up the middleware chain providing the values as you go along
+    duration(durationFromTech) {
+      if (areWeCurrentlyPlayingAnAd(durationFromTech)) {
+        // since we're now in an ad, return the ad duration
+        return 30;
+      } else {
+        // we're playing back content, so, return that duration
+        return 5 * 60;
+      }
+    }
+  }
+});
+```
+
+A simple but interesting middleware to look at is the [playbackrate-adjuster][pra].
+The middleware will change the times of the controls depending on the current rate.
+If you're playing back a 20 minute video and change the rate to 2x, the controls will adjust to display 10 minutes.
+Let's take a look at the code.
+```js
+videojs.use('*', function(player) {
+  /* ... */
+
+  return {
+    setSource(srcObj, next) {
+      next(null, srcObj);
+    },
+
+    duration(dur) {
+      return dur / player.playbackRate();
+    },
+
+    /* ... */
+  };
+});
+```
+So, here, we attach a star-middleware because we want to have it applied to all our videos.
+In `setSource`, we also call `next` directly will `null` and the `srcObj` because we can handle any source.
+We also set up our `duration` method to take in the duration from the previous middleware and divide it by the playback rate we get from the player.
+
+If you look at the [code][] you can see some other methods next to duration.
+They're there to make sure other methods that rely on timing get updated.
+The two methods to notice are `currentTime` and `setCurrentTime`.
+`currentTime` gets called when we want to know what the current time is.
+`setCurrentTime` is called when we're seeking.
+Because the user is seeking in the shifted time, we want to apply our change operation in reverse.
+Instead of dividing it, we want to multiply it.
+```js
+    currentTime(ct) {
+      return ct / player.playbackRate();
+    },
+
+    setCurrentTime(ct) {
+      return ct * player.playbackRate();
+    },
+```
+
+If you were to apply what we've done so far, you'll notice that the nothing changes, the control bar is still showing a duration of 20 minutes.
+This is because as far as Video.js knows, nothing has changed.
+So, we need to tell Video.js that the duration has changed.
+We can do that by storing the tech that Video.js gives us after source selection is complete.
+```js
+videojs.use('*', function(player) {
+  let tech;
+
+  return {
+    setTech(newTech) {
+      tech = newTech;
+    }
+
+    /* ... */
+  };
+});
+```
+
+And then, when the `ratechange` event triggers, we tell Video.js that the duration has changed:
```

Comment:
"and video.js will update the controls accordingly". It may also be good to link to the code that handles this in Video.js